Refactor athena load lambda to use shared paths module

MatMoore opened the pull request:
This refactors the athena load lambda to use the new path construction module in our python base image. (see https://github.com/ministryofjustice/data-platform/pull/1384)

The goal of this refactor is that all the path construction logic is in one place so it's easier to change.

In addition to adding some unit tests, I manually tested this against the following scenarios:

#### Curated table exists, partition files exist
Preconditions:
- the curated table is present in AWS glue, e.g. example_prison_data_product.adjudications
- the data is partitioned and queryable via athena, e.g. "SELECT * FROM "example_prison_data_product"."adjudications" limit 10;"

Expected result:
- the lambda logs "partition for extraction_timestamp and table already exists so nothing more to be done"
- no table left over in data_products_raw database in AWS glue
#### Curated table exists, partition files don't exist
Preconditions
- the curated table is present in AWS glue, e.g. example_prison_data_product.adjudications
- there are no partition files in the curated_data directory, e.g. `s3://data-platform-products-development20230412155038755500000001/curated_data/database_name=example_prison_data_product/table_name=adjudications/extraction_timestamp=20230609T152430Z/`

Expected result:
- lambda logs "table does already exist but partition for timestamp does not"
- the curated table is still present in AWS glue, e.g. example_prison_data_product.adjudications
- no table left over in data_products_raw database in AWS glue
- the partitioned files now exist in the curated_data directory
- the data is partitioned and queryable via athena, e.g. "SELECT * FROM "example_prison_data_product"."adjudications" limit 10;"
#### Curated table doesn't exist, but partition files do exist
Preconditions:
- the curated table is not present in AWS glue, e.g. example_prison_data_product.adjudications
- there are partition files in the curated_data directory, e.g. `s3://data-platform-products-development20230412155038755500000001/curated_data/database_name=example_prison_data_product/table_name=adjudications/extraction_timestamp=20230609T152430Z/`

Expected result:
- the lambda logs "partition data exists but glue table does not"
- no table left over in data_products_raw database in AWS
- the data is partitioned and queryable via athena, e.g. "SELECT * FROM "example_prison_data_product"."adjudications" limit 10;"

#### Curated table doesn't exist, neither do partition files
Preconditions:
- the curated table is not present in AWS glue, e.g. example_prison_data_product.adjudications
- there are no partition files in the curated_data directory, e.g. `s3://data-platform-products-development20230412155038755500000001/curated_data/database_name=example_prison_data_product/table_name=adjudications/extraction_timestamp=20230609T152430Z/`

Expected result:
- the lambda logs "table and partition do not exist but other curated data do"
- no table left over in data_products_raw database in AWS
- the data is partitioned and queryable via athena, e.g. "SELECT * FROM "example_prison_data_product"."adjudications" limit 10;"

murdo-moj commented on containers/daap-athena-load/src/var/task/athena_load_handler.py:
```diff
@@ -28,29 +27,37 @@
 
+    glue_client = boto3.client("glue")
+    athena_client = boto3.client("athena")
+    s3_client = boto3.client("s3")
```
I think a few of these are defined at the beginning of the file

MatMoore replied:
Yeah, they used to be, but I moved them into the handler so that their instantiation during the lifecycle of the tests (otherwise some stubbing broke).

I've since discovered a few different ways of handling import issues here: http://docs.getmoto.org/en/latest/docs/getting_started.html#what-about-those-pesky-imports

So if you prefer I can move this back and try and workaround this another way. Not sure if there is any other advantage/disadvantage to instantiating at module level vs handler level.

murdo-moj replied:
leave it like this by all means, but remove the doubles from the beginning of the file eg athena and s3 on L10/11?

MatMoore replied:
whoops - I didn't spot those duplicates. I've changed this now so they're defined at the module level but can be overridden by tests.

MatMoore commented on containers/daap-athena-load/src/var/task/create_curated_athena_table.py:
```diff
@@ -1,33 +1,32 @@
     metadata,
     logger: DataPlatformLogger,
     s3_security_opts,
+    athena_client,
```
Not sure if we need to add type hints for S3 clients. What do you think? I've left it for now because it seems like it requires adding a 3rd party library https://pypi.org/project/boto3-stubs/

murdo-moj replied:
yeah just leave it if it's another lib

MatMoore commented on containers/daap-athena-load/src/var/task/create_raw_athena_table.py:
```diff
@@ -13,18 +13,17 @@ def create_raw_athena_table(
     table_name = metadata_glue["TableInput"]["Name"]
     try:
-        glue_client.delete_table(DatabaseName="data_products_raw", Name=table_name)
+        glue_client.delete_table(DatabaseName=database_name, Name=table_name)
```
This change is broken because when inferring the schema, we use the curated database name, but the raw table name

MatMoore committed:
Add unit tests to daap-athena-load

Make sure that

- The tests run in CI
- The linter doesn't bug us about repeated code in tests

MatMoore committed:
Add __eq__ method

MatMoore committed:
Remove special path logic from infer_glue_schema

Note there is a slight change of behaviour here.

Before, the database name was based on a parameter, and when called with raw
data, it was being set to the curated database name, (i.e. the one derived from
the data product name). Now, we ask the data_product_config for both the
database name and the table name, which will return a value like
('data_products_raw', 'adjudications_raw'). This behaviour is more consistent,
but we'll need to make sure we handle it correctly in the calling code.

MatMoore committed:
Remove special path logic from create_raw_athena_table

Note that now `infer_metadata_schema` returns the correct database name for raw
data, we can pull the value straight from the metadata.

MatMoore committed:
Remove special path logic from create_curated_athena_table

I've changed the create_curated_athena_table function to accept all the boto
clients as parameters, to make it easier to test.

We're now unit testing some (but not all) of the functionality of
create_curated_athena_table:

- creating the glue db if it is missing
- checking if any partition files exist
- formatting athena queries
- the function doesn't error if partition files already exist
- the function doesn't error if the glue table already exists

Unfortunately we can't test executing the athena queries,
since moto does not support this.

MatMoore committed:
Update athena_load_handler to pass DataProductConfig

This completes the work to remove all code to do with paths and table name
conventions from this lambda.

MatMoore commented on containers/daap-athena-load/tests/unit/athena_load_handler_test.py:
```diff
@@ -0,0 +1,70 @@
+        fake_context,
+    )
+
+    mock_create_curated.assert_called_once_with(
```
Not sure if these assertions are worth keeping - I added them early in the development, but I don't think they provide much value now. We can just test the function runs to completion without raising an error.

murdo-moj replied:
okay, remove them then

murdo-moj commented on containers/daap-athena-load/tests/unit/infer_glue_schema_test.py:
```diff
@@ -0,0 +1,194 @@
+        file_type="parquet",
+    )
+
+    assert metadata_glue == {
```
to make the test files a bit more readable I guess you could test the datatypes bit alone like 
`assert metadata_glue["TableInput"]["StorageDescriptor"]["Columns]`

Emterry reviewed (approved):
ltgm
